### PR TITLE
cargo.eclass: enable trivial crate replacement

### PR DESCRIPTION
**Pull request by Kangie**
Updating vulnerable (or otherwise outdated) crates in Rust ebuilds
is painful. Generally speaking, there are 5 options:

- Run `cargo update` to fetch new versions from the web.
  This is obviously not suitable for use in Portage.
- Patch the software via Portage to accept a non-vulnerable crate.
  This is a reasonable option when the package is not too complex
  but still requires significant developer effort and some familiarity
  with Cargo. In the case of complex patches this may not be feasible,
  or require the generation of a dependency tarball.
- [patch] the source (repository) in Cargo.toml. This enables the
  targeting of specific crates, but does not allow the replacement
  of only a specific version in the depgraph.
- [replace] a particular crate:version in the Cargo.toml. This
  enables the targeting of a particular version with an arbitrary
  path however the replacement crate must *have the same version*
  as the one being overridden.
- `paths = [...]` overrides: pass an array of paths to directories that
  contain a Cargo.toml. Cargo will override any crate with the same package name
  arbitrarily, ignoring the lock file and versions; typically used for testing.
  Is applied via ${CARGO_HOME}/config.toml (i.e. globally)

This Merge Request:

- Implements the `paths` overrides, which will work even when
  Cargo is configured to use a vendored directory. This is not a 'smart'
  replacement and care must be taken to ensure that all versions of
  the crate in use are compatible (`cargo tree` will help).

- Provides a helper which runs `cargo --update --offline` against
  ${ECARGO_VENDOR} (where ${CRATES} are unpacked). This enables the
  replacement of vulnerable versions in ${CRATES}. It is up to the
  consumer to ensure that only the desired crates are being replaced
  and that package behaviour does not change.

- Adds a new `CARGO_BOOTSTRAP` variable which enables packages to
  ignore the minimum version requirement of the eclass. This is only
  used for bootstrapping Rust; if it's being used in any non
  dev-lang/rust ebuilds be sure that you have a good reason.

- Adds `dev-lang/rust{,-bin}:1.54.0`, a step towards bootstrapping
  Rust without needing a binary Rust at all.

Resources:
- https://doc.rust-lang.org/cargo/reference/overriding-dependencies.html
- https://github.com/rust-lang/cargo/issues/3308

---

`app-antivirus/clamav: example of trivial crate replacement` will not be merged; there is no reason to update this package.

There is room for future work that enables us to `[patch]` the offline repository, however since we control it we could just unpack a patched crate there and run `cargo update --offline` which should do effectively the same thing with a lot less effort. We'll wait and see if this is something that consumers of the eclass ask for before putting effort into this; I can only really see it being required in the event that the vulnerable dependent crate isn't receiving updates for some reason.

### CODE REVIEW

**Comment by Kangie** [comment]
Fixed the changes to `_cargo_gen_git_config`, they were caught in a find and replace.

**Comment by whitslack** [comment]
> As the LLVM is far too old for Gentoo we are using the bundled LLVM exclusively.

When I was bootstrapping my Rust toolchain from source, I did resurrect `sys-devel/llvm-13.0.1` and `sys-devel/llvm-14.0.6-r4` from `::gentoo` so I could set `USE="system-llvm"` on `dev-lang/rust` to reduce build times. I grabbed the LLVM ebuilds from the last commits they appeared in before they were pruned. The only change I had to make to get them working again was the rename `sys-devel/make`→`dev-build/make`.

For `dev-lang/rust-1.54.0::gentoo-bootstrap`, there is no `system-llvm` flag available, so I let it build and use its own bundled copy of LLVM 12. I also let `dev-lang/rust-1.55.0::gentoo-bootstrap` build and use its own bundled LLVM 12, even though it _does_ have 'IUSE="system-llvm"`, since it was the only other version that uses LLVM 12.

* `dev-lang/rust-1.54.0::gentoo-bootstrap`: bundled LLVM only
* `dev-lang/rust-1.55.0::gentoo-bootstrap`: LLVM 12 (used only once)
* `dev-lang/rust-1.56.1::gentoo-bootstrap`: LLVM 13
* `dev-lang/rust-1.57.0::gentoo-bootstrap`: LLVM 13
* `dev-lang/rust-1.58.1::gentoo-bootstrap`: LLVM 13
* `dev-lang/rust-1.59.0::gentoo-bootstrap`: LLVM 13
* `dev-lang/rust-1.60.0::gentoo-bootstrap`: LLVM 14
* `dev-lang/rust-1.61.0::gentoo-bootstrap`: LLVM 14
* `dev-lang/rust-1.62.1::gentoo-bootstrap`: LLVM 14
* `dev-lang/rust-1.63.0-r1::gentoo-bootstrap`: LLVM 14
* `dev-lang/rust-1.64.0-r1::gentoo-bootstrap`: LLVM 14
* `dev-lang/rust-1.65.0::gentoo-bootstrap`: LLVM 15
* `dev-lang/rust-1.66.1::gentoo-bootstrap`: LLVM 15
* `dev-lang/rust-1.67.1::gentoo-bootstrap`: LLVM 15
* `dev-lang/rust-1.68.2::gentoo-bootstrap`: LLVM 15
* `dev-lang/rust-1.69.0-r1::gentoo-bootstrap`: LLVM 15
* `dev-lang/rust-1.70.0::gentoo-bootstrap`: LLVM 16
* `dev-lang/rust-1.71.1::gentoo-bootstrap`: LLVM 16
* `dev-lang/rust-1.72.0-r1::gentoo-bootstrap`: LLVM 16
* `dev-lang/rust-1.73.0::gentoo-bootstrap`: LLVM 17
* `dev-lang/rust-1.74.1::gentoo-bootstrap`: LLVM 17
* `dev-lang/rust-1.75.0-r1::gentoo-bootstrap`: LLVM 17
* `dev-lang/rust-1.76.0-r1::gentoo-bootstrap`: LLVM 17
* `dev-lang/rust-1.77.1::gentoo-bootstrap`: LLVM 17
* `dev-lang/rust-1.78.0::gentoo-bootstrap`: LLVM 18
* `dev-lang/rust-1.79.0::gentoo-bootstrap`: LLVM 18
* `dev-lang/rust-1.79.0-r100::gentoo`: LLVM 18
* `dev-lang/rust-1.80.1-r100::gentoo`: LLVM 18
* `dev-lang/rust-1.81.0-r100::gentoo`: LLVM 18
* `dev-lang/rust-1.82.0-r100::gentoo`: LLVM 19

I'm mentioning this here because I was able to save significant time by setting `USE="system-llvm"` while building all those old Rust versions. It was well worth resurrecting `sys-devel/llvm-13.0.1` and `sys-devel/llvm-14.0.6-r4` since they are used by multiple Rust versions. I did not bother to resurrect LLVM 12 since there was only one version of Rust that could have used it, but if you could teach Rust 1.54.0 to use `system-llvm`, then you'd have two versions of Rust that need LLVM 12, and then it might be worth resurrecting LLVM 12 too.

Note that no other packages from the LLVM ecosystem are needed aside from `sys-devel/llvm`. I don't know whether Gentoo will allow the old versions of `sys-devel/llvm` back into the tree, but I will just say that bringing them back into my local overlay did save me some time, and it was easy.

File: eclass/cargo.eclass [hunk 1]
```diff
@@ -37,15 +38,21 @@ case ${EAPI} in
 		if [[ -n ${RUST_MIN_VER} ]]; then
 			# This is _very_ unlikely given that we leverage the rust eclass but just in case cargo requires a newer version
 			# than the oldest in-tree in future.
-			if ver_test "${RUST_MIN_VER}" -lt "${_CARGO_ECLASS_RUST_MIN_VER}"; then
-				die "RUST_MIN_VERSION must be at least ${_CARGO_ECLASS_RUST_MIN_VER}"
+			if [[ -z ${CARGO_BOOTSTRAP} ]]; then
```
Comment (mgorny) [inline eclass/cargo.eclass:41]: This looks really to belong to a separate commit.

File: eclass/cargo.eclass [hunk 2]
```diff
@@ -76,6 +83,39 @@ ECARGO_VENDOR="${ECARGO_HOME}/gentoo"
 # SRC_URI="${CARGO_CRATE_URIS}"
 # @CODE
 
+# @ECLASS_VARIABLE: CRATE_PATHS_OVERRIDE
+# @DEFAULT_UNSET
+# @PRE_INHERIT
+# @DESCRIPTION:
+# Bash string containing paths to crates that will be used to override
```
Comment (mgorny) [inline eclass/cargo.eclass:90]: Err, this doesn't seem to contain paths at all?

File: eclass/cargo.eclass [hunk 3]
```diff
@@ -76,6 +83,39 @@ ECARGO_VENDOR="${ECARGO_HOME}/gentoo"
...
+# `Cargo.toml` in a path will be overridden, ignoring lockfiles,
+# version constraints, etc.
+#
+# This should be used as a last resort where (e.g.) you are
+# bootstrapping Rust and need to override a vendored crate
+# with a newer version, and all versions in use are compatible.
+#
+# Crate names and versions must be separated by a `@`;
+# multiple crates are separated by a space or newline.
+# Crates in CRATE_PATHS_OVERRIDE are implicitly added to CRATES.
```
Comment (mgorny) [inline eclass/cargo.eclass:101]: Does this mean that the replaced crates no longer need to be present in `CRATES`?

File: eclass/cargo.eclass [hunk 4]
```diff
@@ -346,6 +416,39 @@ cargo_target_dir() {
...
+cargo_update_crates () {
+	debug-print-function ${FUNCNAME} "$@"
+
+	if [[ -z ${CARGO} ]]; then
+		die "CARGO is not set; was rust_pkg_setup run?"
+	fi
+
+	local path="${S}/Cargo.toml"
+	if [[ $# -eq 1 ]]; then
+		path="${1}"
```
Comment (mgorny) [inline eclass/cargo.eclass:440]: Any reason not to just `path=${1:-"${S}/Cargo.toml"}`?

File: eclass/cargo.eclass [hunk 5]
```diff
@@ -346,6 +416,39 @@ cargo_target_dir() {
...
+	elif [[ $# -gt 1 ]]; then
+		die "Usage: cargo_update_crates [path_to_Cargo.toml]"
+	fi
+	[[ -f ${path} ]] || die "${path} does not exist"
+
+	set -- "${CARGO}" update --offline --manifest-path ${path}
```
Comment (mgorny) [inline eclass/cargo.eclass:446]: Missing quoting.

File: eclass/cargo.eclass [hunk 6]
```diff
@@ -346,6 +416,39 @@ cargo_target_dir() {
...
 
+# @FUNCTION: cargo_update_crates
+# @USAGE:
+# @DESCRIPTION:
+# Helper function to call `cargo update --offline` with the given Cargo.toml.
+# This will update Cargo.{toml,lock}. This should provide a straightforward
+# approach to updating vulnerable crates in a package.
+#
+# To use: replace any vulnerable crates in ${CRATES} with updated (and compatible)
+# versions, then call `cargo_update_crates` in src_prepare. If Cargo.toml is not
```
Comment (mgorny) [inline eclass/cargo.eclass:427]: I'm confused. Is there a reason why this needs to be called explicitly, when `CRATE_PATHS_OVERRIDE` is pre-inherit?

Comment (Kangie) [reply eclass/cargo.eclass:427]: `CRATE_PATHS_OVERRIDE` is one way of replacing crates. If you do this you put the new crates in `CRATE_PATHS_OVERRIDE` and remove them from `CRATES`. No other action is required as the `Cargo.toml` in the crate we override with will apply to any other crate with the same name regardless of version. We need the additional `CRATE_PATHS_OVERRIDE` variable to tell the eclass which crates need their path added to ${ECARGO_HOME}/.cargo/config.toml`.

However the other method of updating a crate used by an ebuild is to change its version in `CRATES`to a non-vulnerable one, then call `cargo_update_crates` to have the package manager update the lockfile. I couldn't think of a good way of automating this, and there's no need for a separate variable for what's a one-liner either way. Strictly speaking users could implement this themselves in src_prepare; the function is just a helper: "Please update the lockfile against the newer version(s) of crate(s) available in the offline repository masquerading as crates.io"

If you have a suggestion to make this more straightforward I'm happy to hear it!

**Comment by Kangie** [comment]
> > As the LLVM is far too old for Gentoo we are using the bundled LLVM exclusively.

> Note that no other packages from the LLVM ecosystem are needed aside from `sys-devel/llvm`. I don't know whether Gentoo will allow the old versions of `sys-devel/llvm` back into the tree, but I will just say that bringing them back into my local overlay did save me some time, and it was easy.

No, we're not bringing old versions of LLVM back into the tree to save time bootstrapping `dev-lang/rust`. Users interested in the timesave have the option of using `dev-lang/rust-bin` or a binpkg of `dev-lang/rust` to emerge any particular version that is required.

Frankly I'm not sure that we need to cater to saving a little time for users who would rather bootstrap 28 versions of dev-lang/rust between 1.54 and 1.82 to avoid downloading a single binary anyway. :wink:

File: eclass/cargo.eclass (hunk 1 above)
Comment (Kangie) [reply eclass/cargo.eclass:41]: Will extract into its own commit.

File: eclass/cargo.eclass (hunk 3 above)
Comment (Kangie) [reply eclass/cargo.eclass:101]: Yes, that is correct. Because we need to state the crates anyway to know which need a config file insertion ( a bit like `GIT_CRATES` without the associative array), it doesn't make sense to duplicate them in `${CRATES}` when we can infer that the package requires them anyway.

I thought about being even more explicit and saying 'Do not include these in `${CRATES}`', however it shouldn't actually hurt anything if they end up in both.

**Comment by arthurzam** [comment]
> Edit: We will keep the `system-llvm` path (and if possible prefer it) for rusts that need llvm>=16. Ideally `gccrs` will become usable or `mrustc` will get an update that allows bootstrapping a reasonable version of `dev-lang/rust` so that we don't have to worry too much about this but I'm not sure that will come to pass in the short term.

And just to be clear, we aren't promising LLVM 16 stays indefinitely in tree. When LLVM Gentoo project decided to remove LLVM 16, 17, etc' they just disable `system-llvm` for matching rust versions.

LLVM Gentoo project is very understaffed, so we can't add burden to them.

File: eclass/cargo.eclass (hunk 6 above)
Comment (mgorny) [reply eclass/cargo.eclass:427]: Oh, then I got confused that you need to do both.

File: eclass/cargo.eclass (hunk 3 above)
Comment (mgorny) [reply eclass/cargo.eclass:101]: I'm just thinking you could leave a note that they can but do not need to be removed.

**Comment by mgorny** [comment]
Yeah, literally as soon as we can finally remove LLVM 15, I'm starting the tracker for 16.

**Comment by Kangie** [comment]
@mgorny Feedback addressed and comments responded to. No need to respond quickly, it's late enough here that I won't be making more changes tonight!

File: eclass/cargo.eclass (hunk 6 above)
Comment (Kangie) [reply eclass/cargo.eclass:432]: I'll make sure that the 'writing rust ebuilds' wiki page gets an update that clearly explains when to use each of these (and when you'll, unfortunately, still need to patch!)

File: eclass/cargo.eclass (hunk 3 above)
Comment (Kangie) [reply eclass/cargo.eclass]: I've updated the comment, but upon reflection that hasn't actually made the situation any better.

The typical use-case for this is ebuilds that don't already have a CRATES variable anyway (e.g. the rust 1.54 ebuild), as it's typically not something packages mix with vendored crates (that I've seen anyway). I'll ensure that the 'How to write a Rust ebuild' docs spell it out though.
